src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.DecodeRecipients.cs

```diff
@@ -0,0 +1,306 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Pkcs;
+using System.Security.Cryptography.X509Certificates;
+using System.Security.Cryptography.Xml;
+
+namespace Internal.Cryptography.Pal.OpenSsl
+{
+    internal sealed partial class DecryptorPalOpenSsl : DecryptorPal
+    {
+        private static RecipientInfoCollection GetRecipients(DerSequenceReader encodedCms)
+        {
+            DerSequenceReader recipientSet = encodedCms.ReadSet();
+            List<RecipientInfo> recipientInfoStack = new List<RecipientInfo>();
+
+            // RecipientInfo objects are defined as (in RFC 2630) 
+            //
+            // RecipientInfo ::= CHOICE {
+            //      ktri KeyTransRecipientInfo
+            //      kari [1] KeyAgreementRecipientInfo
+            //      kekri [2] KEKRecipientInfo }
+            //
+            // New types have been added but we don't support them
+
+            const byte constructedSequence = (byte)DerSequenceReader.DerTag.Sequence | DerSequenceReader.ConstructedFlag;
+            const byte contextSpecific1 = DerSequenceReader.ConstructedFlag | DerSequenceReader.ContextSpecificTagFlag | 0x01;
+
+            while (recipientSet.HasData)
+            {
+                byte tag = recipientSet.PeekTag();
+                
+                switch (tag)
+                {
+                    case constructedSequence:
+                        KeyTransRecipientInfo ktri = ReadKeyTrans(recipientSet);
+                        recipientInfoStack.Add(ktri);
+                        break;
+                    case contextSpecific1:
+                        recipientInfoStack.AddRange(ReadKeyAgree(recipientSet));
```

Comment:
Rather than creating many temporary lists, consider passing recipientInfoStack into ReadKeyAgree, so it can just add the objects while creating them.